### PR TITLE
[FLINK-7943] Make ParameterTool thread safe

### PR DESCRIPTION
**Pull request by tillrohrmann**
## What is the purpose of the change

This commit changes the serialization of the ParameterTool such that only the
data map is contained. The defaultData and the unrequestedParameters maps are
not serialized because they are only used on the client side. Additionally, the
defaultData and unrequestedParameters map are being made thread safe by using
ConcurrentHashMaps.

## Verifying this change

- Added `ParameterToolTest#testConcurrentExecutionConfigSerialization`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

**Comment by zentol** [comment]
Travis failure:
```
Tests run: 17, Failures: 0, Errors: 6, Skipped: 0, Time elapsed: 0.029 sec <<< FAILURE! - in org.apache.flink.api.java.utils.RequiredParametersTest
testApplyToWithOptionWithLongAndShortNameAndDefaultValue(org.apache.flink.api.java.utils.RequiredParametersTest)  Time elapsed: 0.012 sec  <<< ERROR!
java.lang.UnsupportedOperationException: null
	at java.util.Collections$UnmodifiableMap.put(Collections.java:1457)
	at org.apache.flink.api.java.utils.RequiredParameters.hasNoDefaultValueAndNoValuePassedOnAlternativeName(RequiredParameters.java:147)
	at org.apache.flink.api.java.utils.RequiredParameters.checkAndApplyDefaultValue(RequiredParameters.java:116)
	at org.apache.flink.api.java.utils.RequiredParameters.applyTo(RequiredParameters.java:94)
	at org.apache.flink.api.java.utils.RequiredParametersTest.testApplyToWithOptionWithLongAndShortNameAndDefaultValue(RequiredParametersTest.java:195)

testApplyToWithMultipleTypes(org.apache.flink.api.java.utils.RequiredParametersTest)  Time elapsed: 0.002 sec  <<< ERROR!
java.lang.UnsupportedOperationException: null
	at java.util.Collections$UnmodifiableMap.put(Collections.java:1457)
	at org.apache.flink.api.java.utils.RequiredParameters.hasNoDefaultValueAndNoValuePassedOnAlternativeName(RequiredParameters.java:147)
	at org.apache.flink.api.java.utils.RequiredParameters.applyTo(RequiredParameters.java:103)
	at org.apache.flink.api.java.utils.RequiredParametersTest.testApplyToWithMultipleTypes(RequiredParametersTest.java:228)

testApplyToWithOptionMultipleOptionsAndOneDefaultValue(org.apache.flink.api.java.utils.RequiredParametersTest)  Time elapsed: 0.001 sec  <<< ERROR!
java.lang.UnsupportedOperationException: null
	at java.util.Collections$UnmodifiableMap.put(Collections.java:1457)
	at org.apache.flink.api.java.utils.RequiredParameters.hasNoDefaultValueAndNoValuePassedOnAlternativeName(RequiredParameters.java:147)
	at org.apache.flink.api.java.utils.RequiredParameters.applyTo(RequiredParameters.java:103)
	at org.apache.flink.api.java.utils.RequiredParametersTest.testApplyToWithOptionMultipleOptionsAndOneDefaultValue(RequiredParametersTest.java:210)

testApplyToWithOptionAndDefaultValue(org.apache.flink.api.java.utils.RequiredParametersTest)  Time elapsed: 0 sec  <<< ERROR!
java.lang.UnsupportedOperationException: null
	at java.util.Collections$UnmodifiableMap.put(Collections.java:1457)
	at org.apache.flink.api.java.utils.RequiredParameters.hasNoDefaultValueAndNoValuePassedOnAlternativeName(RequiredParameters.java:147)
	at org.apache.flink.api.java.utils.RequiredParameters.checkAndApplyDefaultValue(RequiredParameters.java:116)
	at org.apache.flink.api.java.utils.RequiredParameters.applyTo(RequiredParameters.java:94)
	at org.apache.flink.api.java.utils.RequiredParametersTest.testApplyToWithOptionAndDefaultValue(RequiredParametersTest.java:182)

testApplyToMovesValuePassedOnShortNameToLongNameIfLongNameIsUndefined(org.apache.flink.api.java.utils.RequiredParametersTest)  Time elapsed: 0.001 sec  <<< ERROR!
java.lang.UnsupportedOperationException: null
	at java.util.Collections$UnmodifiableMap.put(Collections.java:1457)
	at org.apache.flink.api.java.utils.RequiredParameters.hasNoDefaultValueAndNoValuePassedOnAlternativeName(RequiredParameters.java:144)
	at org.apache.flink.api.java.utils.RequiredParameters.applyTo(RequiredParameters.java:103)
	at org.apache.flink.api.java.utils.RequiredParametersTest.testApplyToMovesValuePassedOnShortNameToLongNameIfLongNameIsUndefined(RequiredParametersTest.java:127)

testDefaultValueDoesNotOverrideValuePassedOnShortKeyIfLongKeyIsNotPassedButPresent(org.apache.flink.api.java.utils.RequiredParametersTest)  Time elapsed: 0.001 sec  <<< ERROR!
java.lang.UnsupportedOperationException: null
	at java.util.Collections$UnmodifiableMap.put(Collections.java:1457)
	at org.apache.flink.api.java.utils.RequiredParameters.hasNoDefaultValueAndNoValuePassedOnAlternativeName(RequiredParameters.java:144)
	at org.apache.flink.api.java.utils.RequiredParameters.checkAndApplyDefaultValue(RequiredParameters.java:116)
	at org.apache.flink.api.java.utils.RequiredParameters.applyTo(RequiredParameters.java:94)
	at org.apache.flink.api.java.utils.RequiredParametersTest.testDefaultValueDoesNotOverrideValuePassedOnShortKeyIfLongKeyIsNotPassedButPresent(RequiredParametersTest.java:142)
```

File: flink-java/src/main/java/org/apache/flink/api/java/utils/ParameterTool.java [hunk 1]
```diff
@@ -560,9 +591,13 @@ protected Object clone() throws CloneNotSupportedException {
...
-		ParameterTool ret = new ParameterTool(this.data);
-		ret.data.putAll(other.data);
-		ret.unrequestedParameters.addAll(other.unrequestedParameters);
+		Map<String, String> resultData = new HashMap<>(data.size() + other.data.size());
+		resultData.putAll(data);
+		resultData.putAll(other.data);
+
+		ParameterTool ret = new ParameterTool(resultData);
+
+		ret.unrequestedParameters.putAll(other.unrequestedParameters);
```
Comment (zentol) [inline flink-java/src/main/java/org/apache/flink/api/java/utils/ParameterTool.java]: what about `this.unrequestedParameters`?

File: flink-java/src/main/java/org/apache/flink/api/java/utils/ParameterTool.java [hunk 2]
```diff
@@ -210,15 +212,44 @@ public static ParameterTool fromSystemProperties() {
...
 
 	private ParameterTool(Map<String, String> data) {
-		this.data = new HashMap<>(data);
-		this.defaultData = new HashMap<>();
-		this.unrequestedParameters = new HashSet<>(data.keySet());
+		this.data = Collections.unmodifiableMap(new HashMap<>(data));
+
+		this.defaultData = new ConcurrentHashMap<>(data.size());
+
+		this.unrequestedParameters = new ConcurrentHashMap<>(data.size());
```
Comment (zentol) [inline flink-java/src/main/java/org/apache/flink/api/java/utils/ParameterTool.java]: why not `... = Collections.setFromMap(new ConcurrentHashMap<>(data.size());`

File: flink-java/src/test/java/org/apache/flink/api/java/utils/ParameterToolTest.java [hunk 3]
```diff
@@ -454,124 +469,196 @@ public void testUnrequestedDoubleWithMissingValue() {
...
 			parameter.getUnrequestedParameters());
 
-		Assert.assertEquals(4.0, parameter.getFloat("float"), 0.00001);
-		Assert.assertEquals(createHashSet("double", "string"),
+		assertEquals(4.0, parameter.getFloat("float"), 0.00001);
+		assertEquals(createHashSet("double", "string"),
 			parameter.getUnrequestedParameters());
 
-		Assert.assertEquals(8.0, parameter.getDouble("double"), 0.00001);
-		Assert.assertEquals(createHashSet("string"),
```
Comment (zentol) [inline flink-java/src/test/java/org/apache/flink/api/java/utils/ParameterToolTest.java]: unnecessary changes.

File: flink-java/src/main/java/org/apache/flink/api/java/utils/ParameterTool.java (hunk 2 above)
Comment (tillrohrmann) [reply flink-java/src/main/java/org/apache/flink/api/java/utils/ParameterTool.java]: Good idea. Didn't know that this existed.

File: flink-java/src/main/java/org/apache/flink/api/java/utils/ParameterTool.java (hunk 1 above)
Comment (tillrohrmann) [reply flink-java/src/main/java/org/apache/flink/api/java/utils/ParameterTool.java]: Forgot to add it.

File: flink-java/src/test/java/org/apache/flink/api/java/utils/ParameterToolTest.java (hunk 3 above)
Comment (tillrohrmann) [reply flink-java/src/test/java/org/apache/flink/api/java/utils/ParameterToolTest.java]: True, I guess that was done due to auto reformatting when writing the `testConcurrentExecutionConfigSerialization`.

**Comment by tillrohrmann** [comment]
Thanks for the review @zentol. Yes you're right that this PR has failing test cases. I wasn't aware that the `ParameterTool` is so messy (getters with side effects, storing default values for future gets, counting which parameters have been accessed, sending this whole thing over to the cluster) and that the `RequiredParameters` change the `ParameterTool` directly. I'm currently thinking about letting the `RequiredParameters` create a new `ParameterTool` if they have to change a value. Ideally, I would like to remove `RequiredParameters` and do not send the `ParameterTool` to the cluster.

**Comment by tillrohrmann** [comment]
Fixed the `RequiredParametersTest` by making `RequiredParameters#applyTo` create a new `ParameterTool` whenever changes are needed. Note, however, that this code is not very performant since we copy `ParameterTool` every time we apply a change.

File: flink-java/src/main/java/org/apache/flink/api/java/utils/RequiredParameters.java [hunk 4]
```diff
@@ -83,82 +82,100 @@ public void add(Option option) throws RequiredParametersException {
...
+		// TODO: Throw this all away!!!
+		if (o.hasAlt() && parameterTool.has(o.getAlt())) {
+			HashMap<String, String> newData = new HashMap<>(parameterTool.toMap());
+			newData.put(o.getName(), parameterTool.get(o.getAlt()));
+
+			return ParameterTool.fromMap(newData);
 		} else {
 			if (o.hasDefaultValue()) {
-				data.put(o.getName(), o.getDefaultValue());
+				HashMap<String, String> newData = new HashMap<>(parameterTool.toMap());
```
Comment (zentol) [inline flink-java/src/main/java/org/apache/flink/api/java/utils/RequiredParameters.java]: could we not instead create a copy at the start of applyTo and modify that?

Comment (tillrohrmann) [reply flink-java/src/main/java/org/apache/flink/api/java/utils/RequiredParameters.java]: Yes we could. But I didn't want to touch more than necessary. I think that this actually dead code and should be removed.

**Comment by tillrohrmann** [comment]
Thanks for the review @zentol. I've changed the `RequiredParameters#applyTo` method as you suggested and it was a less impacting change.

File: flink-java/src/main/java/org/apache/flink/api/java/utils/ParameterTool.java (hunk 2 above)
Comment (zentol) [reply flink-java/src/main/java/org/apache/flink/api/java/utils/ParameterTool.java]: IIRC you already switched to a `Set`, did you revert something?

Comment (tillrohrmann) [reply flink-java/src/main/java/org/apache/flink/api/java/utils/ParameterTool.java]: Yes I must have changed something. Don't know really what. Will re-address your feedback.

**Comment by tillrohrmann** [comment]
I've addressed your comments @zentol. Additionally I've rebased onto the latest master and did a force push. Sorry for that.

**Comment by bowenli86** [comment]
LGTM +1

**Comment by zentol** [comment]
+1

**Comment by zentol** [comment]
merging.
